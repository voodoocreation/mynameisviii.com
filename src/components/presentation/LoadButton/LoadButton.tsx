import cn from "classnames";
import * as React from "react";

import { isAlmostInViewport, isServer } from "../../../helpers/dom";
import Button from "../Button/Button";

interface IProps {
  className?: string;
  isDisabled?: boolean;
  isLoading?: boolean;
  isScrollLoadEnabled?: boolean;
  onLoad: () => void;
  triggerDistance?: number;
}

export default class LoadButton extends React.Component<IProps> {
  public static defaultProps = {
    isDisabled: false,
    isLoading: false,
    isScrollLoadEnabled: true,
    triggerDistance: 50
  };

  private buttonRef: React.RefObject<Button> = React.createRef();

  public componentWillMount() {
    if (!isServer() && this.props.isScrollLoadEnabled) {
      window.addEventListener("scroll", this.onWindowScroll);
    }
  }

  public componentDidMount() {
    this.tryLoad();
  }

  public componentWillReceiveProps(nextProps: IProps) {
    if (!isServer()) {
      if (!this.props.isScrollLoadEnabled && nextProps.isScrollLoadEnabled) {
        window.addEventListener("scroll", this.onWindowScroll);
      }

      if (this.props.isScrollLoadEnabled && !nextProps.isScrollLoadEnabled) {
        window.removeEventListener("scroll", this.onWindowScroll);
      }
    }
  }

  public componentDidUpdate() {
    this.tryLoad();
  }

  public componentWillUnmount() {
    if (!isServer() && this.props.isScrollLoadEnabled) {
      window.removeEventListener("scroll", this.onWindowScroll);
    }
  }

  public render() {
    const { children, className, isLoading, ...props } = this.props;

    return (
      <Button
        {...props}
        className={cn("LoadButton", className)}
        isLoading={isLoading}
        onClick={this.onClick}
        ref={this.buttonRef}
      >
        {children}
      </Button>
    );
  }

  private tryLoad = () => {
    if (
      this.props.isScrollLoadEnabled &&
      !this.props.isDisabled &&
      !this.props.isLoading &&
      this.buttonRef.current &&
      isAlmostInViewport(
        this.buttonRef.current.buttonNode,
        this.props.triggerDistance
      )
    ) {
      this.props.onLoad();
    }
  };

  private onWindowScroll = () => {
    this.tryLoad();
  };

  private onClick = () => {
    this.props.onLoad();
  };
}
