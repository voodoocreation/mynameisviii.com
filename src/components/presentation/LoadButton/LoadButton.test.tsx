import { mount, render, shallow } from "enzyme";
import * as React from "react";

import LoadButton from "./LoadButton";

const setup = (fn: any, fromTestProps?: any) => {
  const props = {
    className: "TestLoadButton",
    onLoad: jest.fn(),
    ...fromTestProps
  };

  return {
    actual: fn(<LoadButton {...props}>Load button</LoadButton>),
    props
  };
};

const g: any = global;
const screenWidth = 1920;
const screenHeight = 1080;
const addEventListener = g.addEventListener;
const removeEventListener = g.removeEventListener;

describe("[presentation] <LoadButton />", () => {
  beforeEach(() => {
    g.innerWidth = screenWidth;
    g.innerHeight = screenHeight;
    g.isServer = false;
    g.addEventListener = jest.fn((...args) => addEventListener(...args));
    g.removeEventListener = jest.fn((...args) => removeEventListener(...args));
  });

  it("renders correctly with minimum props", () => {
    const { actual } = setup(render);

    expect(actual).toMatchSnapshot();
  });

  it("triggers `onLoad` prop when button is clicked", () => {
    const { actual, props } = setup(shallow, { isScrollLoadEnabled: false });

    actual.simulate("click");

    expect(props.onLoad).toHaveBeenCalledTimes(1);
  });

  it("attempts automatic load on window scroll when the button is almost in view", () => {
    const { actual, props } = setup(mount);

    actual.instance().buttonRef.current.buttonNode = g.mockElement(
      100,
      100,
      0,
      0
    );
    g.dispatchEvent(new UIEvent("scroll"));

    expect(props.onLoad).toHaveBeenCalledTimes(2);
  });

  it("doesn't attempt automatic load on window scroll when the button isn't almost in view", () => {
    const { actual, props } = setup(mount, { isScrollLoadEnabled: false });

    actual.instance().buttonRef.current.buttonNode = g.mockElement(
      100,
      100,
      screenHeight + 50,
      0
    );
    actual.setProps({ isScrollLoadEnabled: true });
    g.dispatchEvent(new UIEvent("scroll"));

    expect(props.onLoad).not.toHaveBeenCalledTimes(1);
  });

  describe("window scroll event binding", () => {
    it("binds when isScrollLoadEnabled=true", () => {
      setup(shallow);

      expect(g.findMockCall(g.addEventListener, "scroll")).toBeDefined();
    });

    it("doesn't bind when isScrollLoadEnabled=false", () => {
      setup(shallow, { isScrollLoadEnabled: false });

      expect(g.findMockCall(g.addEventListener, "scroll")).toBeUndefined();
    });

    it("binds when `isScrollLoadEnabled` changes from false to true", () => {
      const { actual } = setup(shallow, { isScrollLoadEnabled: false });

      expect(g.findMockCall(g.addEventListener, "scroll")).toBeUndefined();
      actual.setProps({ isScrollLoadEnabled: true });
      expect(g.findMockCall(g.addEventListener, "scroll")).toBeDefined();
    });

    it("unbinds when `isScrollLoadEnabled` changes from true to false", () => {
      const { actual } = setup(shallow, { isScrollLoadEnabled: true });

      expect(g.findMockCall(g.addEventListener, "scroll")).toBeDefined();
      actual.setProps({ isScrollLoadEnabled: false });
      expect(g.findMockCall(g.removeEventListener, "scroll")).toBeDefined();
    });

    it("unbinds when component unmounts", () => {
      const { actual } = setup(shallow);

      expect(g.findMockCall(g.addEventListener, "scroll")).toBeDefined();
      actual.unmount();
      expect(g.findMockCall(g.removeEventListener, "scroll")).toBeDefined();
    });

    it("doesn't unbind when component unmounts and isScrollLoadEnabled=false", () => {
      const { actual } = setup(shallow, { isScrollLoadEnabled: false });

      expect(g.findMockCall(g.addEventListener, "scroll")).toBeUndefined();
      actual.unmount();
      expect(g.findMockCall(g.removeEventListener, "scroll")).toBeUndefined();
    });

    it("doesn't try to bind/unbind when rendering on the server", () => {
      g.isServer = true;
      const { actual } = setup(shallow, { isScrollLoadEnabled: false });

      expect(g.findMockCall(g.addEventListener, "scroll")).toBeUndefined();
      actual.setProps({ isScrollLoadEnabled: true });
      expect(g.findMockCall(g.addEventListener, "scroll")).toBeUndefined();
    });
  });
});
