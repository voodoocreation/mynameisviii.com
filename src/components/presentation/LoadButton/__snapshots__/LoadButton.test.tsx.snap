// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <LoadButton /> renders correctly with minimum props 1`] = `
<button
  class="Button LoadButton TestLoadButton"
  type="button"
>
  <span
    class="Button-content"
  >
    Load button
  </span>
</button>
`;
